onnxruntime/core/session/inference_session.cc

```diff
@@ -1044,7 +1047,16 @@ Status InferenceSession::Run(const RunOptions& run_options, const std::vector<st
     // info all execution providers InferenceSession:Run started
     // TODO: only call OnRunStart for all providers in-use
     for (auto& xp : execution_providers_) {
-      ORT_CHECK_AND_SET_RETVAL(xp->OnRunStart());
+      // call OnRunStart and add to exec_providers_to_stop if successful
+      auto start_func = [&xp, &exec_providers_to_stop]() {
```

Comment:
nit: is there any advantage of using a lambda here?

Reply:
As discussed, the lamda allows using ORT_CHECK_AND_SET_RETVAL is more consistent with other code, but the macro itself is confusing. Will look at an alternative to it in a separate PR.